lib/matplotlib/layoutbox.py

```diff
@@ -0,0 +1,727 @@
+# -*- coding: utf-8 -*-
+"""
+
+Conventions:
+
+"constrain_x" means to constrain the variable with either
+another kiwisolver variable, or a float.  i.e. `constrain_width(0.2)`
+will set a constraint that the width has to be 0.2 and this constraint is
+permanent - i.e. it will not be removed if it becomes obsolete.
+
+"edit_x" means to set x to a value (just a float), and that this value can
+change.  So `edit_width(0.2)` will set width to be 0.2, but `edit_width(0.3)`
+will allow it to change to 0.3 later.  Note that these values are still just
+"suggestions" in `kiwisolver` parlance, and could be over-ridden by
+other constrains.
+
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import kiwisolver as kiwi
+import numpy as np
+import matplotlib
+import warnings
+
+
+# renderers can be complicated
+def get_renderer(fig):
+    if fig._cachedRenderer:
+        renderer = fig._cachedRenderer
+    else:
+        canvas = fig.canvas
+        if canvas and hasattr(canvas, "get_renderer"):
+            renderer = canvas.get_renderer()
+        else:
+            # not sure if this can happen
+            # seems to with PDF...
+            #warnings.warn("constrained_layout : falling back to Agg renderer")
```

Comment:
I have some plan to make canvases always have get_renderer (so that this becomes unnecessary).  In any case, decide whether you want the warning or not :)